crates/interfaces/src/p2p/full_block.rs

```diff
@@ -192,28 +235,307 @@ where
+                                // received bad response
+                                this.client.report_bad_message(peer);
+                            } else {
+                                // sort headers from highest to lowest block number
+                                headers.sort_unstable_by_key(|h| Reverse(h.number));
+
+                                // check the starting hash
+                                if headers[0].hash() != this.hash {
+                                    // received bad response
+                                    this.client.report_bad_message(peer);
+                                } else {
+                                    // get the bodies request so it can be polled later
+                                    let hashes =
+                                        headers.iter().map(|h| h.hash()).collect::<Vec<_>>();
+
+                                    // set the actual request if it hasn't been started yet
+                                    if !this.has_bodies_request_started() {
+                                        this.request.bodies =
+                                            Some(this.client.get_block_bodies(hashes));
+                                    }
+
+                                    // set the headers response
+                                    this.headers = Some(headers);
+                                }
+                            }
+                        }
+                        Err(err) => {
+                            debug!(target: "downloaders", %err, ?this.hash, "Header range download failed");
+                        }
+                    }
+
+                    if this.headers.is_none() {
+                        // received bad response, retry
+                        this.request.headers = Some(this.client.get_headers(HeadersRequest {
+                            start: this.hash.into(),
+                            limit: this.count,
+                            direction: HeadersDirection::Falling,
+                        }));
+                    }
+                }
+                // This branch handles block body responses from peers - it first checks that the
+                // number of bodies matches what we requested.
+                //
+                // If the number of bodies doesn't match, we penalize the peer and retry the
+                // request.
+                // If the number of bodies does match, we assemble the bodies with the headers
+                // received by a previous response, and return the result.
+                RangeResponseResult::Body(res) => {
+                    match res {
+                        Ok(bodies_resp) => {
+                            let (peer, bodies) = bodies_resp.split();
+                            if bodies.len() != this.count as usize {
+                                // received bad response
+                                this.client.report_bad_message(peer);
```

Comment:
we shouldn't penalize here because it's possible that the blocks don't fit in a single packet,

instead this must resubmit the a request for missing hashes